Heiko Code Review 2

CodeLtDave opened the pull request.

CodeLtDave committed:
aktueller Stand kinda working

asda

CodeLtDave committed:
Revert "aktueller Stand kinda working"

This reverts commit 78cfb44d6532ee927103c8c98c154a441fe86af8.

CodeLtDave committed:
MakeStatic files init

Commit:
Almost working

Commit:
kleiner Schubser

CodeLtDave committed:
Working Rewrite but other errors

CodeLtDave committed:
Revert "Working Rewrite but other errors"

This reverts commit 9a56081415a75f1029d56aaea3674a292a739c18.

Commit:
right before change working

Commit:
working

Commit:
HEUREKA

CodeLtDave committed:
fix public modifier disappearing

leave the remeaining modifiers instead of replacing them

CodeLtDave committed:
beautify

CodeLtDave committed:
cleanup

CodeLtDave committed:
Revert "cleanup"

This reverts commit 8c021e42514c000b4628f2b5b16a949f11b9ed18.

Commit:
InvocationUpdate added but not working

CodeLtDave committed:
Chage Invocations, work in progress

Commit:
Working Invocation change in the same class

Commit:
Refactoring now working in different classes

MethodInvocations will now also be changed in different classes within the same project. In other projects its not working properly.

Commit:
Bugfix

Bugfix for Refactoring of more than one method in the same class that individually contain a Methodinvocation

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

Commit:
Bugfix

Fixed Bug when refactoring a method with another return type than void. Now the Refactoring works for all return types.

Commit:
Changed MethodInvocations don't overwrite arguments anymore

Commit:
First MakeStaticRefactoring Test implemented

Commit:
First Test Working

CodeLtDave committed:
better code structure and cleanup

Commit:
More Tests added

Commit:
Additional Tests and Bugfix for Refactoring

If MethodDeclaration is modified before MethodInvocations, the Refactoring will fail

Commit:
ThrowException Test working

CodeLtDave committed:
added ChangeManager which handels adding edits

Before there could be multiple CompilationUnitChanges for one CompilationUnit. Now you can use

addEditToChangeManager(TextEdit, ICompilationUnit)

so that this is automatically take care off. If there is a CompilationUnitChange already its pulled from ChangeManager and edited instead of having multiple CompilationUnitChanges on one ICompilationUnit.

CodeLtDave committed:
cleanup

CodeLtDave committed:
small fix

CodeLtDave committed:
small refactorings

Commit:
new tests

new tests for every type as parameter and return type

Commit:
new test for this-usage in MethodDeclaration

Commit:
Additional Tests for multiple refactored files

CodeLtDave committed:
Improve method invocation handling in MakeStaticRefactoring

Enhance the MakeStaticRefactoring class to better handle method
invocations in both instance and static contexts. Adjustments
include passing 'this' as the new argument for invocations within
instance methods and passing the instance variable for invocations
in other classes or static methods.

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

Commit:
small change in test

Commit:
Additional test for this + InnerClass

Commit:
small fix for test

CodeLtDave committed:
Implement handling of instance usages in method refactoring now with a CHECK

This is not finished yet, there are 2 tests failing but at least no errors anymore

In this commit, we introduced the fHasInstanceUsages flag to track if a method being refactored has any instance variables or methods. If fHasInstanceUsages is true, the method will be modified to accept an additional parameter representing the instance of the class.

Key changes include:

Implementing an ASTVisitor to visit SimpleName nodes within the method body and check for instance variables or methods
If an instance variable is found, the code replaces the variable with an appropriate FieldAccess expression
If an instance method is found, the code replaces the method invocation with an appropriate MethodInvocation expression

This update improves the refactoring process by correctly handling methods with instance usages, enabling a more accurate static method conversion.

Commit:
Beautify tests and proper suggestion of refactoring

Based on the specific selection of the user the refactoring will now only be suggested if the user selects a method.

Commit:
Additional functionality and adjustments

Additional checks for Refactoring in checkInitialConditions. Editor Selection for Refactoring is now more precisely.
Adjustments for tests for more accurate selection.
Refactoring can now be executed from outline window.
Message Constants included for better Exception and RefactoringStatus handling.

Commit:
Adjustments to tests

Commit:
New Tests and handling for SuperMethodInvocations

Commit:
Little fix for tests that expect error

Commit:
Implemented ASTMatcher for Tests

The method matchASTs is more precise and ignores eventual whitespaces in the files. With the method matchFiles it is easier to understand where the files have differences.  Both methods can be used for testing

Commit:
Check if parameter with same name already exists

if Method has instance usage a new parameter is introduced to the method. If this parameter has the same name as an already existing parameter then an error will be thrown.

Commit:
Check if duplicate method signature exists

when refactoring a method with instance usage, a new parameter is introduced. After the refactoring the method could have the same signature as an already existing method. This check throws an Error in this case.

Commit:
Bugfix Overlapping TextEdits when status has error

Commit:
Multiple Updates

- Better Handling of SuperMethodInvocations.
- Deletion of @Override Annotation.
- Better Handling of StatusErrors -> immediate return after error

Commit:
ShortCut implemented: Alt+Shift+K

Commit:
Bugfix for box name on Refactoring errors

Commit:
MakeStatic can now be used from upper Refactoring menu

Commit:
Cleanup

Commit:
Check if selected method is already static + Bugfix

- new test for this case
- new error message
- fixed a bug where an "Overlapping text edits" error was thrown when pressing back in WIzard after an error occured and then again pressing continue to continue the refactoring.
- Changed a line, so that MakeStatic error message and not IntroduceIndirection error  message will be thrown in MakeStaticAction

Commit:
Adjusted Handling of SuperMethodInvocations and methods that are overriden in TypeHierarchy

Commit:
Code Cleanup

- deleted method getCachedCURewrite()
- more comments for better understanding
- fMethodDeclaration is not initialized in checkInitialConditions() and not in findMethodDeclaration()
- findMethodDeclaration() is now called checkForInstanceUsagesInMethodDeclaration() and is only for this use.

Commit:
New Tests for special cases

- Generic declaration
- Non Static Inner Class

CodeLtDave committed:
tests working

CodeLtDave committed:
Merge branch '_DavidNEW' into _MichaelNEW

Commit:
Additional tests and handling

Tests for following cases:
- Usage of Generics
- refactoring a method should not add additional parameter in case of field access whenparameter already exists
- Selected method is overriding a method of parent type and has no parameter -> refactoring would change method so that it hides parent type

Also included handling for last case

Commit:
Additional Tests

new tests for following cases:
- Field access in anonymous class of a field that belongs to the outer class
- The same case but in a lambda and not in an anonymous class

CodeLtDave committed:
removed findMethodDeclaration and added functionality to new inner class"ChangeInstanceUsagesInMethodBody"

Commit:
New Tests

- Many new tests for various use cases

Commit:
Better handling of duplicate method signature

Before method were only compared with number of parameters and  method name. Now also the parameter types are compared

Commit:
Better handling for duplicate parameterName

Commit:
Handling for static method in lower hierarchy

- handling added if method with same signature in lower hierarchy is static -> refactoring only throws error if method in lower hierarchy overrides selected method, but a static method can not override.
- Bug fixed for test

Commit:
Adjustments to tests, Little change to input page, Bugifx

- Fixed a bug where methods that override parent method and have instance usage couldnt be refactored
- Adjustments to tests so they work properly
- Deleted not needed tests
- Little Change to Input Page and Wizard

Commit:
Added warning when refactoring method that overrides

- Changed dialog window size back to default

Commit:
implemented check for recursive method

- User gets error message when refactoring a recursive method.
- This behavior does not apply for SuperMethodinvocations. Methods that override the parent method and call it with super can still be refactored.

Commit:
Added functionality for updating javadocs

Commit:
Adjustment to ExtraTests

Commit:
Added functionality for Generics

- Adjustments to tests
- methods with generics can now be refactored but only if they are not already parametrized with type parameters

Commit:
Updated Generic functionality

- methods that use less generic parameters than the amount of type parameters of the class will now be refactored correctly. Only the exact amount of needed parameters will be added to the type parameter list of the method.
- New Tests for special cases with generics added

Commit:
bug fix if method has array as parameter

Commit:
New Test

Commit:
Several bugfixes for generics functionality

- additional test

Commit:
new Generics functionality

- added functionality for already parametrized methods
- added functionality for bounds of typeParameters

Commit:
little fix for test

- parametrized methods can be refactored now, so test doesnt need to check for error

Commit:
check for WildCardTypes as bounds implemented

- additional Tests and check for wildCardTypes as bounds of typeparam of class. (T extends List<?>)

Commit:
Bugfix and adjustment to tests

- Refactoring is now possible if a method does not override the parent type method but has the same name as the parent type method

Commit:
Bugfix for generics in JavaDoc

- now braces <T> are also added to JavaDoc when adding type parameters

Commit:
check for LambdaExpression and MethodReference

- if Lambda is detected as MethodInvocation the refactoring is cancelled
- bugfix for invoking refactoring on MethodInvocation that is in different class than MethodDeclaration
- code cleanup

Commit:
Functionality for VarArgs included

- new parameter for instance access is now inserted as first parameter of method, so that varargs are always the last parameter
- changed every test that adds new parameter to new behavior

Commit:
Cleanup

- extracted methods from method modifyMethodDeclaration() for more clarity

Commit:
little change

CodeLtDave committed:
fix: hasInstanceUsage and InnerClassCreatation

fix for wrongfully recognised hasInstanceUsages and creating anonymous inner instance classes

Commit:
Bugfix for generics

- updateTypeParamList now also includes typeParams of parametrized methodParams

Commit:
small changes to some tests

Commit:
changes to Tests

CodeLtDave committed:
unfinished state before extracting handleInstanceVariable

Commit:
merging of test classes and changes for MethodReferences

- Changed behaviour for MethodReferences and SuperMethodReferences
- LambdaExpressions are now allowed for Refactoring
- Merging of test classes

Commit:
New tests

- It is not possible anymore to refactor a method that uses explicit super method invocations in the methodDeclaration body
- New warning for explicit access of fields from the parent type with the super keyword
-

Commit:
Comparing of methods is now more precise

Commit:
new test for foncatenated field access

CodeLtDave committed:
extend testConcatenatedFieldAccessAndQualifiedNames

CodeLtDave committed:
fixed testConcatenatedFieldAccessAndQualifiedNames

CodeLtDave committed:
fixes for modifyMethodInvocation and concatinated FieldAccess/QualifiedName

CodeLtDave committed:
fixes for modifyMethodInvocation

CodeLtDave committed:
testsWorking, fixed behavior for InnerFieldAccessInAnonymousClass

Commit:
Deleted output files for tests with error...

-  check if source is available for selected method
- New test for this case

CodeLtDave committed:
checks for anonymous/inner classes

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

Commit:
right output files deleted and others inserted again

CodeLtDave committed:
fixed test22?

CodeLtDave committed:
Letzter SCHUBSSER

Commit:
Merging of tests and new test for nested classes

Commit:
Test merging completed

CodeLtDave committed:
fixed testNested

CodeLtDave committed:
comment refactorings to modifyMethodDeclaration

CodeLtDave committed:
refactorings, broke testInheritance3

CodeLtDave committed:
structure fix to checkNode

Commit:
fixed bug for testInheritance3

Commit:
error messages now better to understand

Commit:
deleted obsolete test testNoAdditionalParameter2

Commit:
Put all "then" statements into a block for readability. Autoformated code

CodeLtDave committed:
Revert accidental change in file

Commit:
improved readability for test class

CodeLtDave committed:
Move MakeStaticAction to org.eclipse.jdt.ui\ui\org\eclipse\jdt\ui\actions

CodeLtDave committed:
Move InputPage and Wizard

CodeLtDave committed:
Move MakeStaticRefactoring.java

CodeLtDave committed:
Revert "Move MakeStaticRefactoring.java"

This reverts commit 6005e5706ab3510f5f5dcc03988d85f3a1fe779c.

Commit:
changes for better understanding

- changed dialog title
- outsourced strings to localized variable

CodeLtDave committed:
Move MakeStaticRefactoring.java

CodeLtDave committed:
Merge branch '_DavidNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _DavidNEW

CodeLtDave committed:
MakeStaticWizard removed unused imports

CodeLtDave committed:
Added "&k" symbol for acessing the action via keypress

CodeLtDave committed:
adjusted copyright notice MakeStaticAction

CodeLtDave committed:
Merge branch '_DavidNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _DavidNEW

CodeLtDave committed:
Modified all Copyright notices

CodeLtDave committed:
changed AST version for parser to IASTSharedValues.SHARED_AST_LEVEL

CodeLtDave committed:
java docs edit, "if...return" restructure

Commit:
Improved code readability for inner class ChangeInstanceUsagesInMethodBody

Commit:
changed inner class ChangeInstanceUsagesInMethodBody to a standalone class

Commit:
deleted unnecessary line of code

Commit:
CleanUp MakeStaticInputPage

Commit:
new names for variables in InputPage

Commit:
small fix to Exception and deleted unnecessary null check

Commit:
JavaDocs for MakeStaticWizard

Commit:
name change for ChangeInstanceUsageInMethodBody

CodeLtDave committed:
create package for MakeStaticRefactoring.java, class split preperation

Commit:
JavaDocs for InstanceUsageRewriter

Commit:
JavaDiocs for MakeStaticRefactoring

Commit:
JavaDocs for fields in MakeStaticRefactoring

Commit:
Outsourced inner class MethodReferenceFinder and simplified its methods

Commit:
JavaDocs for MakeStaticRefactoring methods

CodeLtDave committed:
init ContextCalculator

Commit:
Merged methods isOverridden and is Overriding

Commit:
added field fStatus and changed all usages of status

Commit:
Copyright comment for classes

CodeLtDave committed:
progress

Commit:
improved readability for method modifyMethodInvocation

Commit:
tests for javaDocs

CodeLtDave committed:
unfinished

Commit:
improved readability

Commit:
almost working

Commit:
added general initial condition checks

Commit:
bugfix

Commit:
deleted throw

we make a daily now

Commit:
bugfix for selection of MethodInvocation before starting refactoring

Commit:
all tests working but bad code style

CodeLtDave committed:
logic fix and java docs

CodeLtDave committed:
Unified code paths for TextSelection and IMethod inputs

Unified code paths by calculating IMethod for text selection input, reducing code complexity and improving maintainability

CodeLtDave committed:
refactorings and java docs

CodeLtDave committed:
Merge branch '_DavidNEW' into _MichaelNEW

CodeLtDave committed:
merge JavaDocs and fStatus field

Commit:
small refactorings

and changing back method isOverriding to its initial state because the combination of isOverridden and isOverriding makes no sense anymore.

Commit:
small refactorings and created new class FinalConditionsChecker

Commit:
Outsourced checks

Every check from method checkFinalConditions is now checked from the class FinalConditionsChecker.

Only checks from classes MethodReferenceFinder and InstanceUsageRewriter are not outsourced yet.

Commit:
outsourced checks of MethodReferenceFinder to FinalConditionsChecker

Commit:
changed InitialConditionsChecker methods to static

Commit:
JavaDocs for InititalConditionsChecker

Commit:
fixed copyright message

Commit:
JavaDocs and copyright note for FinalConditionsChecker

Commit:
additional copyright notes

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,599 @@
+ * @since 3.29
+ *
+ */
+public class MakeStaticRefactoring extends Refactoring {
```
You have reduced the visibility of all but this class within the `makestatic` package to only make this class visible outside of the plugin.
That is great in terms of isolation, still there is also another option that I would personally prefer here:
This class could still be placed in the parent package `...refactoring.code` and only use the classes inside this package as utilities (which then have to be made public). Then there is no need for exporting the `makestatic` package, simplifying the plugin's API.

CodeLtDave replied:
Done.
15a7b3f58807b623d13217539a6178f54ec7b1bc

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,599 @@
+	}
+
+	@Override
+	public RefactoringStatus checkInitialConditions(IProgressMonitor pm) throws CoreException, OperationCanceledException {
```
The class became much better to read by extracting the condition checks out of the refactoring. Now it is quite easy to read through this method to see which checks and calculation are performed.

I still have some proposals for improvement. Currently, this method serves as a kind of mediator between the context calculator and the condition checker. This methods has to keep track of the proper state of the context calculation progress while at the same time performing the according checks. The method has to know which calculations have to be performed on the context to be able to retrieve the appropriate values to pass to the condition checker. You could to the following to further decouple them:
1. The condition checkers know which calculations have to be performed in order to provide some property via a getter method. So you could instead having `get` and `calculate` method, you could have `getOrCompute` methods that return the value if it already exists and compute it otherwise. Then it becomes the sole responsibility of the context calculator to know about the process of calculation for providing the according values.
2. Instead of passing specific arguments to static methods of the condition checker, you could make the context calculator non-static and pass it to the condition checker. Then, whenever you call one of its check methods, the check method itself can retrieve the information from the calculation context. And the calculation context in turn knows how to compute that value, if it does not have it stored yet.

As a consequence, this method would only only operate on the condition checker, calling its methods one after another and processing potential fails.

CodeLtDave replied:
Done like discussed 9296ee882f176a7eef52d9fdf8d75ea8053cb8ad

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,599 @@
+
+	@Override
+	public RefactoringStatus checkFinalConditions(IProgressMonitor progressMonitor) throws CoreException, OperationCanceledException {
+		modifyMethodDeclaration();
```
It would be great to have everything related to the final condition checking (i.e., `modifyMethodDeclaration`) separated into another class as well, just like the initial condition checks. Basically, this cal may be replaced with something like `fStatus = new FinalConditionChecker(/* necessary arguments */).checkConditions()` and all the private helper methods are moved to that class.

We can also talk about that one offline to find out, what impediments may exist to perform that refactoring.

HeikoKlare replied:
I just saw that there already is some `FinalConditionChecker` that is called at some placed. Maybe you can do it the same way as for the initial conditions and the context: have a `ChangeCalculator` class and a `FinalConditionChecker` class, move everything related to the change calculation to the former class, which you pass to the final condition checker. Then this method would only call the check methods of the final condition checker, which, in turn, retrieves the required values form the change calculator. The change calculator either has the values or calculates them. Finally, the change calculator can be reused by this refactoring class to provide the final change, just by some `getOrComputeChange` method.

This would also resolve the "weirdness" of just returning a change that was calculated by the condition check on-the-fly, which you have mentioned to me before. When you clearly separate change calculation and condition checking, it becomes explicit that the change is calculates during condition checks "by accident" and that when retrieving the final change from the change calculator it simply returns a cached value (which you do not see from outside, as you could also have called the `getOrComputeChange` method without performing the condition checks before.

CodeLtDave replied:
@Michael5601 introduced the ChangeCalculator in d45f79f12201f4cad1062af666091a8848425a53

Michael5601 replied:
Done.

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ContextCalculator.java:
```diff
@@ -0,0 +1,217 @@
+		} catch (JavaModelException e) {
+			System.err.println("Failed to get the source range of the method: " + iMethod.getElementName()); //$NON-NLS-1$
+			System.err.println("Status: " + e.getJavaModelStatus()); //$NON-NLS-1$
+			e.printStackTrace();
```
You should avoid UI interactions within a class that is only responsible for business logic. This exception should be return to the caller to further process it (either by rethrowing the exception or by returning some error). Then the caller can decide how to present that to the user (e.g. in a dialog or the like). If you return `null` here, the calling method will throw a `NullPointerException`, won't it? That would be pretty bad, because the user gets some logging in his or her console and at the same time some popup giving him a `NullPointerException` while the refactoring overall fails.

CodeLtDave replied:
Done.
43efe702b5750a5ab27efe82edf685cb65b7fcde
39cf9bb7244c318daffdeed35f7c0bdaab32ad8a

CodeLtDave replied:
In these 2 commits i changed the behavior so that instead of printing to the console an error gets thrown. This made me consider other parts of the code that right now are creating possible exceptions.

**Should high-level methods such as checkInitialConditions, checkFinalConditions, and createChange should throw exceptions, I propose an alternative approach.**

Instead of propagating exceptions upward, we could handle them immediately where they occur. This involves surrounding the potentially problematic calls with a try-catch block. If an exception arises, we can then create and return the NotAvailableOnThisSelection Error.

This approach enables us to manage exceptions more locally and avoid throwing exceptions from methods that play a central role in our application logic. This also prevents the propagation of exceptions that may be challenging for the callers to handle appropriately.

**On the downside handling the exception locally makes it harder to find the error if it occurs. While propagating the exceptions gives some insight using the console and stack trace, just returning the NotAvailableOnThisSelection Error does not make it very obvious what went wrong.**

CodeLtDave replied:
![image](https://github.com/CodeLtDave/eclipse.jdt.ui/assets/42917573/4aac7f05-3892-4a13-85c6-05f6897c525c)
Reference which methods cause checkFinalConditions to include the ``throws CoreException``

HeikoKlare replied:
Since the `checkInitialConditions`, `checkFinalConditions` etc. have to handle `CoreExceptions` anyway, you can also let your `check...` methods throw the exceptions instead of catching them (and printing something to the console).

CodeLtDave replied:
Ok e378c0cd6822f918802e84bd602d5bd61903929a
66f37564a27d410ac297899cc1ef392207a71add

CodeLtDave committed:
introduce concept of getOrCompute in ContextCalculator

CodeLtDave committed:
Rethrow JavaModelException in business logic layer

This commit addresses an issue where a JavaModelException was being caught
and handled within a class that should only be responsible for business logic.

Instead of handling the exception and logging it within the method, the exception
is now thrown back to the caller. This allows the caller to decide how best to handle
the error, whether that be rethrowing the exception, returning an error, or presenting
the error to the user in a UI dialog.

This change also prevents potential NullPointerExceptions that could occur when the
method returns a null value. These NullPointerExceptions could cause the refactoring
process to fail and present the user with both console logging and a popup error message.

The changes made align the class more closely with the principle of separation of concerns,
keeping business logic and UI interaction separate.

CodeLtDave committed:
Rethrow JavaModelException in business logic layer 2

Commit:
Init ChangeCalculator

CodeLtDave committed:
missing import JavaModelException

CodeLtDave committed:
Restructure makestatic package for improved isolation

Michael5601 committed:
changed name of method modifyMethodDeclaration

Michael5601 committed:
addInstanceAsParamIfUsed is now executed before the two checks beneath it

Michael5601 committed:
deleted duplicate description in javaDocs of method getName

Michael5601 committed:
JavaDocs for MakeStaticRefactoring

Michael5601 committed:
JavaDocs and removing unnecessary fields in ChangeCalculator

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,249 @@
+	 * The {@code IMethod} object representing the selected method on which the refactoring should
+	 * be performed.
+	 */
+	private IMethod fTargetMethod;
```
I would propose to have the "cached" values `fTargetMethod`, `fTargetMethodDeclaration` etc. only represented in the `fContextCalculator`.
This results in slightly longer code due to more `getOrCompute...` calls, but it ensure that no one will rely on these fields being set at some point in time when it is actually not guaranteed. Always calling `getOrCompute` moves the responsibility for ensuring that the value is present to the context calculator.

Michael5601 replied:
Done except for `MethodDeclaration`. 6061d068c3

Michael5601 replied:
Done by @CodeLtDave. 3f2ea3a00e

Michael5601 replied:
Done by @CodeLtDave. 3f2ea3a00e

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,249 @@
+
+		if (selectionInputType == SelectionInputType.TEXT_SELECTION) {
+			fStatus.merge(InitialConditionsChecker.checkTextSelectionStart(fContextCalculator.getSelectionEditorText()));
+			fStatus.merge(InitialConditionsChecker.checkValidICompilationUnit(fContextCalculator.getSelectionICompilationUnit()));
+			if (fStatus.hasError()) {
+				return fStatus;
+			}
```
Each individual check is still quite verbose and the code looks quite repetitive.
I would make the following suggestion: Make the `InitialConditionsChecker` and `FinalConditionsChecker` instantiable classes, which accept a `RefactoringStatus` as constructor input. Then all the `check...` methods can merge the results into the given status on their own, i.e., `fStatus.merge(...)` can happen within the `check...` methods, and return whether they found an error. Then the caller (i.e., this method) can instantiate the checker `InitialConditionsChecker checker = new InitialConditionsChecker(fStatus)` and simply call `if (!checker.checkValid...) { return fStatus }` for each check. That would make the code here even more concise and slightly easier to read.

Michael5601 replied:
Done for checkInititalConditions. See 5bbee980e3 and 311f6537fd.
For checkFinalConditions the other changes you mentioned should be incorporated first.

Michael5601 replied:
I dont think it is good to change the FinalConditionsChecker to an instantiable class like the InitialConditionschecker. There are many classes that use the static methods of the FinalConditionsChecker and i would need to pass it to every constructor of these classes. Its more concise and clear how it is now. Lets discuss this in a meeting please.

Michael5601 replied:
Done as discussed in the meeting and it worked great. It is even more concise now and I could eliminate some of the last checks that were not inside FinalConditionChecker. Now there is no return of RefactoringStatus in any method except for checkFInalConditions and checkInitialConditions. Status is now only a field in the checkers and in the MakeStaticRefactoring class.

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/MakeStaticRefactoring.java:
```diff
@@ -0,0 +1,249 @@
+		}
+
+
+		fTargetMethod= fContextCalculator.getOrComputeTargetIMethod(); //TODO remove those which dont have to necessarily be fields
```
is that todo still valid?

Michael5601 replied:
Nope and this line is also deleted already. Change is in 6061d068c3.

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ContextCalculator.java:
```diff
@@ -0,0 +1,224 @@
+	 * corresponding {@link IMethodBinding} from that instance.
+	 */
+	private void calculateTargetIMethodBinding() {
+		if (getOrComputeSelectionASTNode() instanceof MethodInvocation selectionMethodInvocation) {
+			fTargetIMethodBinding= selectionMethodInvocation.resolveMethodBinding();
+		} else if (fSelectionASTNode instanceof MethodDeclaration selectionMethodDeclaration) {
+			fTargetIMethodBinding= selectionMethodDeclaration.resolveBinding();
+		}
```
As far as I understand the code, the `calculate...` methods are always expected to return something that is not `null` by their callers (i.e., `getOrComputed...`. Otherwise anyone using the result of `getOrCompute` may retrieve a `null` value that then run into an NPE. For example, calling `calculateTargetIMethod()` will throw an NPE in case the selection ast node is neither a `MethodInvocation` nor a `MethodDeclaration`. So there should always be a fallback logic in these cases.

CodeLtDave replied:
Done.
d0c8e158cf6c9df594271bea787c28e260880317
e8b0bd9886075306e61107d583305c2c0eab270a

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ContextCalculator.java:
```diff
@@ -0,0 +1,224 @@
+	private void calculateTargetIMethodBinding() {
+		if (getOrComputeSelectionASTNode() instanceof MethodInvocation selectionMethodInvocation) {
+			fTargetIMethodBinding= selectionMethodInvocation.resolveMethodBinding();
+		} else if (fSelectionASTNode instanceof MethodDeclaration selectionMethodDeclaration) {
```
Also use `getOrComputeSelectionASTNode` instead of `fSelectioNASTNode` here.

Michael5601 replied:
Done. 3ec669b895

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ContextCalculator.java:
```diff
@@ -0,0 +1,224 @@
+
+	/**
+	 * Resolves the method declaration and binding for the target method.
+	 * @throws JavaModelException
```
Throws documentation must not be empty

Michael5601 replied:
Done. 4136532269

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ChangeCalculator.java:
```diff
@@ -0,0 +1,502 @@
+		while (true) {
+			for (SingleVariableDeclaration param : parameters) {
+				String paramString= param.getName().getIdentifier();
+				if (!(combinedName.equals(paramString))) {
+					duplicateExists= false;
+				} else {
+					duplicateExists= true;
+					break;
+				}
```
```suggestion
				duplicateExists = combinedName.equals(paramString);
				if (duplicateExists) {
					break;
				}
```

Michael5601 replied:
I could not commit this change directly from here so I made the commit myself. ae34115fda

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ChangeCalculator.java:
```diff
@@ -0,0 +1,502 @@
+		boolean duplicateExists= false;
+		String combinedName= classNameFirstLowerCase;
+		int counter= 2;
+		while (true) {
```
You can use `do { ... } while (duplicateExists)` instead of a `while(true)` with some return inside.

Michael5601 replied:
Done but I have the feeling its possible to do it even more concise :D. 562690966f

CodeLtDave replied:
I restructred the method it might even be more concise :D
71927abe3d73f052fc6b523956b7e2437e102285

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/ChangeCalculator.java:
```diff
@@ -0,0 +1,502 @@
+		}
+		boolean duplicateExists= false;
+		String combinedName= classNameFirstLowerCase;
+		int counter= 2;
```
Call this `incrementingSuffix` or something else more comprehensible than `counter`.

Michael5601 replied:
Done. c9c8745555

HeikoKlare commented on org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/refactoring/code/makestatic/FinalConditionsChecker.java:
```diff
@@ -0,0 +1,151 @@
+		paramTypesOfSelectedMethodExtended[0]= extendedClassName;
+		String[] paramTypesOfSelectedMethod= imethod.getParameterTypes();
+
+		for (int i= 0; i < paramTypesOfSelectedMethod.length; i++) {
```
Use a more speaking name like `paramNumber`.

Michael5601 replied:
Done. 604b3341c4

HeikoKlare reviewed (commented):
I really like how the `MakeStaticRefactoring` class looks now an how the actual functionality has been assigned to appropriate and rather self-contained classes.

The problem now is that the error-tracking logic for the final condition checks is spread over `FinalConditionChecker`, `ChangeCalculator` and `MakeStaticRefactoring`. I completely understand why it cannot be extracted from `ChangeCalculator` completely, as it is deep within the individual steps. And it is totally fine for me to keep it there. However, I would then propose to completely integrate the checks into the change calculation, also the few ones that can be (and are currently) extracted out of it.
Reason: Currently, if you read the `MakeStaticRefactoring.checkFinalConditions`, it looks like that method explicitly calls all the necessary check. However, some checks are hidden inside the individual change calculations. I would prefer to have this explicit, so that change calculation takes care of performing all checks. This is, of course, different from how the initial condition checks work, but that totally makes sense, because the initial condition checks just consider the existing, unchanged code, while the final condition checks depend on dynamically calculated changes.

I have also placed some detailed comments. I have not finished an in-depth review for time reasons, but submit this one to give you some timely feedback.

One minor general comment: Please also apply auto-formatting to all added source code files (e.g., by pressing Ctrl + Shift + F in the file), so that the files are formatted uniformly and according to the defined style preferences.

Michael5601 committed:
InititalConditionsChecker is now instantiated

all checks in checkInitialConditions are now more concise and easier to read. The checker methods are instance methods now and return a boolean value instead of the RefactoringStatus. The InititalConditionsChecker is instantiated with a RefactoringStatus object and uses it in every checker method to determine which boolean value needs to be returned.

Michael5601 committed:
bugfix and updated @return tag in JavaDocs

Michael5601 committed:
updated description for @throws tag in JavaDocs

Michael5601 committed:
refactor code in ChangeCalculator.

Michael5601 committed:
cleaner solution for code in generateUniqueParameterName

while (true) is changed to do while

Michael5601 committed:
changed variable name of counter to incrementingSuffix

Michael5601 committed:
reduced instance variables

Deleted instance variables in MakeStaticRefactoring and replaced the usage with getOrCompute method invocations.

Michael5601 committed:
changed the direct usage of an instance variable to the matching get method invocation

Michael5601 committed:
changed name of index i to parameterNumber

Michael5601 committed:
description for exception

Michael5601 committed:
moved final condition checks from MakeStaticRefactoring into ChangeCalculator

CodeLtDave committed:
remove try catch

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

CodeLtDave committed:
removed try catch blocks

CodeLtDave committed:
removed fTargetMethodDeclaration field

Michael5601 committed:
modifications to MethodDeclaration are capsulated in modifyMethodDeclaration

- Changecalculator has new field fInstanceUsageRewriter and field fTargetMethodhasInstanceUsage is delted instead. Everywhere fTargetMethodhasInstanceUsage was used the getter getTargetMethodhasInstanceUsage from InstanceUsageRewriter is used

CodeLtDave committed:
Refactor generateUniqueParameterName method

    - Improved variable naming for better readability.
    - Removed redundant null check on method parameters.
    - Streamlined loop logic to improve performance.
    - Created a separate method `parameterNameExists` to handle the concern of checking parameter name existence.

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

Michael5601 committed:
JavaDocs for modifyMethodDeclaration

Michael5601 committed:
getChanges is now getOrComputeChanges

getOrComputeChanges checks if changes are present and if not modifyMethodDeclaration is performed before returning allchanges

Michael5601 committed:
addition to javadocs

CodeLtDave committed:
added Copyright JavaDoc to ChangeCalculator

CodeLtDave committed:
removed outdated ToDo

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

CodeLtDave committed:
added fallback logic to ContextCalculator for Exceptions

CodeLtDave committed:
added fallback logic to ContextCalculator for Exceptions 2

CodeLtDave committed:
small fix

Michael5601 committed:
The RefactoringStatus is now fully managed by FinalConditionsChecker and InitialConditionsChecker

- all checks of checkFinalConditions take place in FinalConditionsChecker now. No RefactoringStatus is returned by any method except checkFInalConditions and checkInitialConditions

Michael5601 committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

CodeLtDave committed:
introduced progess monitor

Michael5601 committed:
JavaDocs for MethodReferenceFinder

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

CodeLtDave committed:
auto format

CodeLtDave committed:
Merge branch '_MichaelNEW' of https://github.com/CodeLtDave/eclipse.jdt.ui into _MichaelNEW

CodeLtDave committed:
Merge branch 'master' into _MichaelNEW

Michael5601 committed:
JavaDocs for InstanceUsageRewriter

Michael5601 committed:
JavaDocs for FinalConditionsChecker

Michael5601 committed:
JavaDocs for ChangeCalculator

CodeLtDave committed:
adjustements to fork sync

CodeLtDave committed:
Convert IProgressMonitor to SubMonitor

CodeLtDave committed:
add checkCancel for SubMonitor
